### PR TITLE
Editor: Make asset registration .asset.php optional for blocks

### PR DESCRIPTION
**Pull request by josephfusco**
It is a hard requirement that a *.asset.php file is present to register a block in WordPress. Otherwise, it throws a Warning:

> function register_block_script_handle was called incorrectly. The asset file (/wp-content/uploads/faustwp/blocks/block-a/index.asset.php) for the "editorScript" defined in "create-block/block-a" block definition is missing. (This message was added in version 5.5.0.)

and the block will not appear in the inserter.

Trac ticket: https://core.trac.wordpress.org/ticket/60460

Gutenberg issue: https://github.com/WordPress/gutenberg/issues/57234

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by gziolo** [comment]
Resharing my https://github.com/WordPress/gutenberg/issues/57234#issuecomment-1875059772 from the issue in the Gutenberg repository. I propose that we remove the handling for the missing asset file and provide the default values instead. To land this patch, the required changes would be removing the proposed filter, and the `_doing_it_wrong` handling. It would be great to add a unit test that covers the case when there is no asset file provided.

**Comment by colorful-tones** [comment]
It looks like unit tests are failing. I do not have the knowledge to fix or write 'em. Would it be necessary for merging, and could we have a separate follow-up PR to address the tests @gziolo ?

**Comment by colorful-tones** [comment]
Also, would it be easier to move this patch to be against the Gutenberg repo?

**Comment by josephfusco** [comment]
A lot of the failing tests were fixed with https://github.com/WordPress/wordpress-develop/pull/5799/commits/5c67d336cc762757a2123e0e6c0f982efd52d12c, however there are still a few tests failing but seem unrelated.

> Also, would it be easier to move this patch to be against the Gutenberg repo?

If the code for `register_block_script_handle` is sourced from outside of wordpress-develop, I'm happy to move this patch there.

File: src/wp-includes/blocks.php [hunk 1]
```diff
@@ -185,7 +175,16 @@ function register_block_script_handle( $metadata, $field_name, $index = 0 ) {
 		$script_args['strategy'] = 'defer';
 	}
 
-	$script_asset        = require $script_asset_path;
+	if ( file_exists( $script_asset_path ) ) {
+		$script_asset = require $script_asset_path;
+	} else {
+		// Ensure graceful fallback for missing or inaccessible *.asset.php by setting default dependencies and version.
+		$script_asset = array(
```
Comment (gziolo) [inline src/wp-includes/blocks.php]: Would it be possible to simplify it to:

```suggestion
		$script_asset = array();
```

given that access to the array is guarded later and the same default values seem to be provided?

File: tests/phpunit/tests/blocks/register.php [hunk 2]
```diff
@@ -220,7 +220,6 @@ public function test_wrong_array_index_do_not_register_block_script_handle() {
 	}
 
 	/**
-	 * @expectedIncorrectUsage register_block_script_handle
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/register.php:223]: It looks like we had a case with a missing asset file covered, so it no longer has to assert for doing it wrong call 👍

File: tests/phpunit/tests/blocks/register.php [hunk 3]
```diff
@@ -234,6 +233,52 @@ public function test_missing_asset_file_register_block_script_handle() {
...
+	public function test_missing_asset_php_for_traditionally_registered_block() {
+		$metadata = array(
+			'file'   => DIR_TESTDATA . '/blocks/notice/block.json',
+			'name'   => 'unit-test/traditional-block',
+			'script' => 'file:./blocks/notice/block.js',
+		);
+		$result   = register_block_script_handle( $metadata, 'script' );
+
+		// Assert that the result is false due to missing .asset.php
+		$this->assertFalse( $result );
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/register.php]: What’s the reason that it returns `false`? Shouldn't it return the generated handle name instead?

**Review by gziolo — COMMENTED** [review]
Thank you for iterations. All changes need to get applied in WordPress core.

File: src/wp-includes/blocks.php [hunk 4]
```diff
@@ -162,18 +162,8 @@ function register_block_script_handle( $metadata, $field_name, $index = 0 ) {
...
-			sprintf(
-				/* translators: 1: Asset file location, 2: Field name, 3: Block name.  */
-				__( 'The asset file (%1$s) for the "%2$s" defined in "%3$s" block definition is missing.' ),
-				$script_asset_raw_path,
-				$field_name,
-				$metadata['name']
-			),
-			'5.5.0'
-		);
 		return false;
```
Comment (gziolo) [inline src/wp-includes/blocks.php]: Ok, we should remove this early exit so the asset gets correctly registered.

File: src/wp-includes/blocks.php [hunk 5]
```diff
@@ -185,7 +175,16 @@ function register_block_script_handle( $metadata, $field_name, $index = 0 ) {
 		$script_args['strategy'] = 'defer';
 	}
 
-	$script_asset        = require $script_asset_path;
+	if ( file_exists( $script_asset_path ) ) {
```
Comment (spacedmonkey) [inline src/wp-includes/blocks.php]: Can we avoid this file exists? This check is bad for performance.

Comment (spacedmonkey) [reply src/wp-includes/blocks.php]: Could we try an include instead?

**Review by spacedmonkey — CHANGES_REQUESTED** [review]

**Comment by colorful-tones** [comment]
Nice progress here @josephfusco ! 🕺

File: tests/phpunit/tests/blocks/register.php [hunk 6]
```diff
@@ -258,6 +243,21 @@ public function test_handles_passed_register_block_script_handles() {
...
+	 */
+	public function test_missing_asset_file_register_block_script_handle_with_default_settings() {
+		$metadata = array(
+			'file'   => __FILE__,
+			'name'   => 'unit-tests/test-block',
+			'script' => 'file:./blocks/notice/missing-asset.js',
+		);
+		$result   = register_block_script_handle( $metadata, 'script' );
+
+		$this->assertSame( 'unit-tests-test-block-script' );
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/register.php]:
```suggestion
		$this->assertSame( 'unit-tests-test-block-script', $result );
```

File: src/wp-includes/blocks.php [hunk 7]
```diff
@@ -185,7 +170,7 @@ function register_block_script_handle( $metadata, $field_name, $index = 0 ) {
 		$script_args['strategy'] = 'defer';
 	}
 
-	$script_asset        = require $script_asset_path;
+	$script_asset        = @include $script_asset_path;
```
Comment (gziolo) [inline src/wp-includes/blocks.php]:
```suggestion
	// Asset file for blocks is optional. See https://core.trac.wordpress.org/ticket/60460.
	$script_asset        = @include $script_asset_path;
```

**Review by gziolo — APPROVED** [review]
Updating code directly on GitHub isn't the best idea after all 😅 

The good news is that it's ready to go!

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/57559.

**Comment by colorful-tones** [comment]
Thanks for getting this over the line @gziolo and thanks @josephfusco for kicking it off 🙌

**Pull request closed without merging**